### PR TITLE
Adding Gamma poisson factorization

### PR DESCRIPTION
**Pull request by alexis-cvetkov**
Changes:

- Added `gamma_poisson_factorization.py` which implements online Gamma-Poisson factorization for encoding string variables.
- Added the corresponding tests in`test_gamma_poisson_factorization`.
- Modified examples 02 and 03 to include this method.
- Updated `CHANGES.rst` and `index.rst` to describe this new method.

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
Wohoo!!

@Phaide : can you have a look when you have time?

Have a look at the code, the test (it's OK if you don't understand everything, this is a sophisticated mathematical algorithm. Check that you understand how to use this, that it feels natural, that the examples make sense, that it's well referenced in the docs (we should point from SimilarityEncoder to this for large datasets). Thanks!!!!

File: dirty_cat/gamma_poisson_factorization.py
Comment (Phaide) [inline dirty_cat/gamma_poisson_factorization.py:84]: Wouldn't it be ``analyzer`` instead of ``analizer`` ?

Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:84]: You're right, I'll fix this !

**Comment by alexis-cvetkov** [comment]
Following the error on travis, now I check for the sklearn version when importing the "k-means++" init method from `sklean.cluster.k_means_`, since its name has changed in the development version.

**Review by Phaide — COMMENTED** [review]
Very good !
Thanks @alexis-cvetkov, good work !

The code is clean and fairly documented (though a few more inline comments to explain complex operations would be welcome!).
One last request: as mentioned by Gaël, could you please consider adding a little notice in the ``SimilarityEncoder`` docstring ?
Thanks again !

**Comment by amy12xx** [comment]
Great to see this addition. I look forward to going through the code when I have a bit more time.

**Comment by alexis-cvetkov** [comment]
Thank you for the review @Phaide, I have added more documentation in the code for gamma-poisson factorization, and a little docstring in `similarity_encoder.py`.
This PR looks ready for merge to me now !

**Review by Phaide — APPROVED** [review]
Looks good to me !
@amy12xx could you please have a quick look before merging ?
Thanks !

File: CHANGES.rst
Comment (amy12xx) [inline CHANGES.rst:7]: Typo in intrerpretable

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:11]: Is it better to say " to maximize their likelihood on the data" instead?

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:20]: It may be better to only include the modules we need instead of import sklearn

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:64]: If we are keeping to sklearn's naming conventions, then we can use bool instead of boolean.

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:186]: Using a list comprehension will be faster.

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:213]: Should we initialize in the init method instead?

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:218]: I guess we can do away with the else block

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:318]: We can use a different name since iter is a python keyword.

File: dirty_cat/gamma_poisson_factorization.py
Comment (amy12xx) [inline dirty_cat/gamma_poisson_factorization.py:504]: Is there a reason to pass B? Right now it does not seem to be used.

**Review by amy12xx — COMMENTED** [review]
This looks great @alexis-cvetkov . I left a few minor comments.

File: dirty_cat/test/test_gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/test/test_gamma_poisson_factorization.py:40]:
```suggestion


```
PEP8 asks for 2 empty lines between first-level functions or classes.

File: dirty_cat/test/test_gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/test/test_gamma_poisson_factorization.py:50]:
```suggestion


```

PEP8 asks for 2 empty lines between first-level functions or classes.

File: doc/index.rst
Comment (GaelVaroquaux) [inline doc/index.rst:18]:
```suggestion
a fast encoding method.

```

An empty line is required to start a new paragraph.

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [reply dirty_cat/gamma_poisson_factorization.py:20]: +1

Comment (GaelVaroquaux) [reply dirty_cat/gamma_poisson_factorization.py:20]: I actually think that this line is not necessary.

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:41]: I would like to suggest to rename this class to a more marketting-friendly name: I fear that the current name will scare users into thinking that this is a highly-technical class, while actually this is really the class that they should be using.

I propose `GapEncoder`. `Gap` is a fairly classic abbreviation of Gamma Poisson.

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:157]: All the lines below should be done in the beginning of fit, and not the init. In scikit-learn style, inits have no other logic than storing parameters in attributes

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:159]: In scikit-learn style, as this attibute is data-dependent, it must be named with a trailing underscore.
```suggestion
            self.ngrams_count_ = HashingVectorizer(
```

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:164]: In scikit-learn style, as this attibute is data-dependent, it must be named with a trailing underscore.
```suggestion
                self.word_count_ = HashingVectorizer(
```

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:169]: In scikit-learn style, as this attibute is data-dependent, it must be named with a trailing underscore.
```suggestion
            self.ngrams_count_ = CountVectorizer(
```

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:172]: In scikit-learn style, as this attibute is data-dependent, it must be named with a trailing underscore.
```suggestion
                self.word_count_ = CountVectorizer()
```

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:44]: We need to give a bit of intuition here. The technical details can go in the "notes" section that goes at the end of the docstring.

For the top, I suggest something like:
```
This encoder can be understand as a continuous encoding on a set of latent categories estimated from the data.
The latent categories are built by capturing combinations of substrings that frequently co-occur.
```

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [inline dirty_cat/gamma_poisson_factorization.py:49]: I know that everybody is going to hate me, but should we be using a different term than "topic". The term "topic" comes from the topic-modeling origin of the Gamma-Poisson Factorization, but here the user probably does not think in terms of topic modeling. Should we use `n_components`, as in the MinHashEncoder?

**Review by GaelVaroquaux — COMMENTED** [review]
Here are a bunch of cosmetic comments.

I'll try to come back and check the algorithmics, but we had checked them back when we coded this originally.

Thanks for the hard work!

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:20]: I agree, I just import the variable sklearn.__version__ now.

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:186]: What do you have in mind ? I cannot use assignments like `self.H_dict[x] = h` in a list comprehension, so something like `[self.H_dict[x] = h for x, h in zip(X, H)]` doesn't work.

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:213]: Yes, I moved it there!

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:218]: Yes, I moved `self.rho_ = self.rho` in the init method.

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:504]: You're right, removed!

File: dirty_cat/gamma_poisson_factorization.py
Comment (GaelVaroquaux) [reply dirty_cat/gamma_poisson_factorization.py:218]: No, it should not be in the init, because if people change rho or batch_size before calling fit, the change won't be propagated. I guess what @amy12xx meant is that you can do:
```python
self.rho_ = rho
if self.rescale_rho:
       # Make update rate per iteration independant of the batch_size
       self.rho_ = self.rho ** (self.batch_size / len(X))
```

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:41]: I agree! Changed into `GapEncoder`.

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py:49]: Yes, this will be consistent with the minhash encoder.

File: CHANGES.rst
Comment (alexis-cvetkov) [reply CHANGES.rst:8]: Here is an attempt:
> Added online Gamma-Poisson factorization through the :class:\`GapEncoder\` class. This method discovers latent categories formed via combinations of substrings, and encodes string data as combinations of these categories. To be used if interpretability is important.

**Comment by GaelVaroquaux** [comment]
> Here is an attempt:

>     Added online Gamma-Poisson factorization through the :class:`GapEncoder`
>     class. This method discovers latent categories formed via combinations of
>     substrings, and encodes string data as combinations of these categories. To
>     be used if interpretability is important.

This is great!!

File: dirty_cat/test/test_gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/test/test_gamma_poisson_factorization.py:16]: This test takes about 40s to run, and less if I decrease the number of samples/components. Should it be done on only a fraction of possible combinations then ?

**Comment by GaelVaroquaux** [comment]
> This test takes about 40s to run, and less if I decrease the number of samples/ components. Should it be done on only a fraction of possible combinations then ?

Yes please. I'm not sure that being exhaustive is crucial. I'd rather
work on a smaller number of well-chosen combinations. I would keep the
number of samples, so that the test is reliable.

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py]: I'd say "likelihood of the data", because we tune H and W so that the observed data is the most plausible.

**Comment by alexis-cvetkov** [comment]
CircleCI is not building anymore. I think this is because it now uses python 3.9 during the build, and some packages like scikit-learn do not support it yet.

Update: I pinned the version of python to 3.8 in `build_doc.sh`. I don't know if it's the right way to fix this but it works!

**Review by amy12xx — COMMENTED** [review]
I had something like this in mind: self.H_dict = {x: h for x, h in zip(X, H)} but I think I assumed that H_dict would have the exact same elements as X. If this is not the case, you can ignore this comment.

File: dirty_cat/gamma_poisson_factorization.py
Comment (alexis-cvetkov) [reply dirty_cat/gamma_poisson_factorization.py]: Actually, I don't know. The code for the GapEncoder was initially implemented by Patricio Cerda and Gael, so I kept the default parameters. However when I fit this encoder to 100,000 input strings, the memory charge is about 500Mb/CPU thread, which is maybe a bit much. Decreasing the batch_size to 100 solves this, and I checked that it doesn't change the performance of GapEncoder on examples 02/03, so I propose to use that instead. Maybe @GaelVaroquaux has an opinion on that ?

**Comment by GaelVaroquaux** [comment]
> Decreasing the batch_size to 100 solves this,

Good idea: try putting it to 128.

Thanks!

**Comment by amy12xx** [comment]
Thanks @alexis-cvetkov . Looks good to me.

**Review by Phaide — APPROVED** [review]

File: dirty_cat/gap_encoder.py
Comment (GaelVaroquaux) [inline dirty_cat/gap_encoder.py:403]: Likewise: why is the partial_fit commented out? It's a very useful feature.

**Review by GaelVaroquaux — CHANGES_REQUESTED** [review]
Thanks a lot for the PR. I finally had time to review it in depth.

I made a bunch of inline comments. On top of these, I have a few bigger picture comments:

- We need to add a comment at the top of the "Scalability considerations" example saying that the GapEncoder is naturally scalable and should be preferred for large-scale data.
- We should restore the get_feature_names method and demo it in one of the examples
- We should restore the partial_fit method and demo it in a new example on scalability

Thanks!!!

File: dirty_cat/gap_encoder.py
Comment (alexis-cvetkov) [reply dirty_cat/gap_encoder.py:160]: Is there a specific reason to initialize `H_dict` in `fit` rather than in `__init__` ? Also, it means that multiple calls of `fit` will update the same `H_dict` and not retrain the model from scratch. Is that what we want ?

**Comment by GaelVaroquaux** [comment]
> Is there a specific reason to initialize H_dict in fit rather than in __init__ ?

It implies that every call to fit has a clean start, independent of the past.

> Also, it means that multiple calls of fit will update the same H_dict and not retrain the model from scratch. Is that what we want ?

I'm confused: isn't it the other way around: initialize in fit means that
multiple calls of fit will not update the same H_dict (partial_fit should
behave differently), no?

**Comment by alexis-cvetkov** [comment]
> It implies that every call to fit has a clean start, independent of the past.

Okay, I was confused by the test you proposed:
```
if not hasattr(self, "H_dict"):
    ...
```
In this case I don't need to test if `H_dict` exist but I can simply re-initialize it everytime `fit` is called

**Comment by GaelVaroquaux** [comment]
> Okay, I was confused by the test you proposed:

OK, this was an error on my side. Thanks for noting it.

**Comment by GaelVaroquaux** [comment]
@alexis-cvetkov : careful, I am pushing to your branch. If you work on this, pull first.

**Comment by GaelVaroquaux** [comment]
I believe that this is now good to go. I will merge when the tests have run. Thanks!!!

**Comment by alexis-cvetkov** [comment]
Great ! Thank you !

**Pull request merged**
